### ISSUE TITLE
analyse_freenrg_mbar is broken

### ISSUE THREAD

**Issue by jmichel80** [michellab/Sire#82]
Using the latest devel version of sire to produce a trajectory, then performing analysis as written below

[ julien@mechagodzilla: ~/projects/SAMPL/reference/na_cyc/small/12A/discharge/output ]$ cat MBAR.sh 
~/sire.app/bin/analyse_freenrg_mbar -i lambda-0.000/simfile.dat lambda-0.125/simfile.dat lambda-0.250/simfile.dat lambda-0.375/simfile.dat lambda-0.500/simfile.dat lambda-0.625/simfile.dat lambda-0.750/simfile.dat lambda-0.875/simfile.dat lambda-1.00/simfile.dat --lam 0.00 0.125 0.250 0.375 0.500 0.625 0.750 0.875 1.00 --temperature 298.0 -o mbar.pmf
[ julien@mechagodzilla: ~/projects/SAMPL/reference/na_cyc/small/12A/discharge/output ]$ bash MBAR.sh 
"/home/julien/sire.app/share/Sire/scripts/analyse_freenrg_mbar.py" 
Starting /home/julien/sire.app/bin/analyse_freenrg_mbar: number of threads equals 16
# Writing all output to file mbar.pmf

We are doing a timeseries analysis using the timeseries analysis module in pymbar and will subsample according to that.
We are doing a timeseries analysis using the timeseries analysis module in pymbar and will subsample according to that.
/home/julien/sire.app/lib/python3.4/site-packages/pymbar/utils.py:70: FutureWarning: comparison to `None` will result in an elementwise object comparison in the future.
  if N_k == None:
K (total states) = 9, total samples = 1467
N_k = 
[ 163.  163.  163.  163.  163.  163.  163.  163.  163.]
There are 9 states with samples.
Initializing free energies to zero.
Initial dimensionless free energies with method zeros
f_k = 
[ 0.  0.  0.  0.  0.  0.  0.  0.  0.]
/home/julien/sire.app/lib/python3.4/site-packages/scipy/optimize/minpack.py:236: RuntimeWarning: xtol=0.000000 is too small, no further improvement in the approximate
  solution is possible.
  warnings.warn(msg, RuntimeWarning)
Final dimensionless free energies
f_k = 
[ 0.         -0.66327738 -1.3632676  -2.09888765 -2.86906772 -3.67300011
 -4.50996649 -5.37941022 -6.28084813]
MBAR initialization complete.
Free energy difference estimated with MBAR is: -3.719440 kcal/mol +/- 0.022837 kcal/mol
Free energy difference estimated with MBAR is: -9.215693 kJ/mol +/- 0.095549 kJ/mol
Free energy difference estimated with TI is: -6.246511 kcal/mol

The conversion from kcal/mol to kJ/mol is wrong. 
The TI result is the BAR result times kT.

**Comment by ppxasjsm** [comment]
Ok strange. The cluster version seems to be fine. 
I am just in the middle of fixing a few things with sire right now anyway.

**Comment by jmichel80** [comment]
Another (minor issue). The tool generates the cryptic error message below if the number of lambda windows passed via the command line does not match the number of foreign energies in the simfile. A more helpful error message or guarantee that the error will not be encountered would be good. 

[ julien@mechagodzilla: ~/projects/SAMPL/reference/na_cyc/small/12A/vanish/output ]$ bash MBAR.sh 
"/home/julien/sire.app/share/Sire/scripts/analyse_freenrg_mbar.py" 
Starting /home/julien/sire.app/bin/analyse_freenrg_mbar: number of threads equals 16

# Writing all output to file mbar.pmf
We are doing a timeseries analysis using the timeseries analysis module in pymbar and will subsample according to that.
Traceback (most recent call last):
  File "/home/julien/sire.app/share/Sire/scripts/analyse_freenrg_mbar.py", line 173, in <module>
    subsample_obj.subsample_energies()
  File "/home/julien/sire.app/lib/python3.4/site-packages/Sire/Tools/FreeEnergyAnalysis.py", line 224, in subsample_energies
    self._subsampled_u_kln[k,:,:] = self._u_kln[k,:,indices_k[k]].transpose()
ValueError: could not broadcast input array from shape (12,197) into shape (11,197)

**Comment by Steboss** [comment]
Did you check if the number of lambdas windows under --lam is the same of the lambda folder given with -i ?
it has happened to me too and that was the explanation

I got another error which is:

# Writing all output to file mbar.dat
Traceback (most recent call last):
  File "/home/steboss/sire.app/share/Sire/scripts/analyse_freenrg_mbar.py", line 155, in <module>
    data = np.array(data)
ValueError: could not broadcast input array from shape (50000,22) into shape (50000)

basically I can't understand why but some simfile.dat have 22 terms rather than 9 and that cause a "mis"shaping

**Comment by Steboss** [comment]
New question:

is there a way to analyse vacuum simualtion with mbar?
I got absolute free energy calculations where in the second step ( q=0 VdW-> 0 ) calcualtions cannot be performed since the gradients are 0.0 kcal/mol at lambda-1and statistical efficacy cannot be computed

**Comment by jmichel80** [comment]
MBAR (the method) doesn't use gradients. It should work. Maybe you need to disable the TI code written by Antonia. It could be buggy if your gradients time series is always 0 kcal/mol because sub sampling will fail. 
It's strange that your gradients are always 0. They could be close to, but should not always be exactly equal to 0 kcal/mol. Sounds like a bug/input file error. Check that your molecule has LJ sites being turned off in the pert file.

**Comment by Steboss** [comment]
My fault
Gradients are not 0 
it's the potential energy at 0.0 kcal/mol, which is rather strange because I am expecting bonded term to be present, is it correct?

**Comment by jmichel80** [comment]
I can reproduce Stefano's error for simulations where the energies stored in simfile are always 0.0000 kcal/mol

Data is available on section6
julien@node009:~/projects/somd/SAMPL/reference/phenol/vac/run001-short/vanish/output$ pwd
/home/julien/projects/somd/SAMPL/reference/phenol/vac/run001-short/vanish/output
julien@node009:~/projects/somd/SAMPL/reference/phenol/vac/run001-short/vanish/output$ cat ../mbar.sh 
#!/bin/bash
#SBATCH -o analyse-free-nrg-%A.%a.out
#SBATCH -p main
#SBATCH -n 1
#SBATCH --time 00:05:00

srun ~/sire.app/bin/analyse_freenrg_mbar -i lambda-0.000/simfile.dat lambda-0.100/simfile.dat lambda-0.200/simfile.dat lambda-0.300/simfile.dat lambda-0.400/simfile.dat lambda-0.500/simfile.dat lambda-0.600/simfile.dat lambda-0.700/simfile.dat lambda-0.800/simfile.dat lambda-0.900/simfile.dat lambda-0.950/simfile.dat lambda-1.000/simfile.dat --lam 0.000 0.100 0.200 0.300 0.400 0.500 0.600 0.700 0.800 0.900 0.950 1.000 --temperature 298.0 -o mbar.pmf > freenrg-MBAR.dat
julien@node009:~/projects/somd/SAMPL/reference/phenol/vac/run001-short/vanish/output$ cat analyse-free-nrg-1053.4294967294.out 
"/home/julien/sire.app/share/Sire/scripts/analyse_freenrg_mbar.py" 
Traceback (most recent call last):
  File "/home/julien/sire.app/share/Sire/scripts/analyse_freenrg_mbar.py", line 173, in <module>
    subsample_obj.subsample_energies()
  File "/home/julien/sire.app/lib/python3.4/site-packages/Sire/Tools/FreeEnergyAnalysis.py", line 209, in subsample_energies
    g_k[i] = timeseries.statisticalInefficiency(self._energies_kn[i,:])
  File "/home/julien/sire.app/lib/python3.4/site-packages/pymbar/timeseries.py", line 164, in statisticalInefficiency
    raise ParameterError('Sample covariance sigma_AB^2 = 0 -- cannot compute statistical inefficiency')
pymbar.utils.ParameterError: Sample covariance sigma_AB^2 = 0 -- cannot compute statistical inefficiency
srun: error: mechagodzilla: task 0: Exited with exit code 1

* We should

1) Fix analyse_free_nrg_mbar so that the code works if the timeseries contain identical energies
2) Establish why the energies in simfile.dat are 0.0 kcal/mol. This is a simulation of a phenol molecule in vacuum without any charges or LJ terms. There should still be some bonded terms from angles/dihedrals/impropers that give a non zero potential energy.

**Comment by ppxasjsm** [comment]
Commit 7e5e3074f445b2ecfa89289b89a26e91badad824 fixed this.

**Comment by SofiaBariami** [comment]
Hi, five years later I came across the same issue even if I have the updated version of Sire with commit 7e5e307. Running vacuum simulations of a small ligand, the vanish leg at lambda-1.00 gives zero energies at simfile.dat. The input files for the simulation are [here](https://github.com/SofiaBariami/sharing_files/tree/master/vav_G1_input) and simfile.dat is [here](https://github.com/SofiaBariami/sharing_files). Did you find out why the energies were zero then?

**Comment by SofiaBariami** [comment]
Hi, five years later I came across the same issue even if I have the updated version of Sire with commit 7e5e307. Running vacuum simulations of a small ligand, the vanish leg at lambda-1.00 gives zero energies at simfile.dat. The input files for the simulation are [here](https://github.com/SofiaBariami/sharing_files/tree/master/vav_G1_input) and simfile.dat is [here](https://github.com/SofiaBariami/sharing_files). Did you find out why the energies were zero then?
